include/boost/spirit/home/x3/operator/kleene.hpp

```diff
@@ -32,7 +35,7 @@ namespace boost { namespace spirit { namespace x3
             while (detail::parse_into_container(
                 this->subject, first, last, context, rcontext, attr))
```

Comment:
Shouldn't `parse_into_container` short circuit the loop on` !has_expectation_failure(context);` ? Or is it already done in this PR somewhere?

Reply:
That's a good question; it does short circuit. There are 3 patterns here:

1. subject parser succeeded (-> returned `true`)
2. subject parser failed normally (-> returned `false`)
3. subject parser failed with expectation failure (-> returned `false`)

So the `while` statement does break when expectation failure is raised. All we need to do in kleene is to distinguish the difference between case 2 and 3, then return false if it was the latter. (Note that case 2 is a legit '0 element match' but case 3 is a solid error.)